const express = require('express');
const bodyParser = require('body-parser');
const Sequilize = require('sequelize');

const { Address, Employee } = require('./src/models');
const { Book, User } = require('./src/models');
const config = require('./src/config/config')
const app = express();

app.use(bodyParser.json());

const sequelize = new Sequilize(config.development);

//para diminuir a complexidade ðŸ – sem arquitetura MSC
app.get('/employees', async (_req, res) => {
  try {
    const employees = await Employee.findAll({
      include: { model: Address, as: 'addresses' },
    });

    return res.status(200).json(employees);
  } catch (e) {
    console.log(e.message);
    res.status(500).json({ message: 'Ocorreu um erro' });
  };
});

//Managed transactions
app.post('/employees', async (req, res) => {
  try {
    const { firstName, lastName, age, city, street, number } = req.body;

    const result = await sequelize.transaction(async (t) => {
      const employee = await Employee.create({
        firstName, lastName, age
      }, { transaction: t });

      await Address.create({
        city, street, number, employeeId: employee.id
      }, { transaction: t });

      return res.status(201).json({ message: 'Cadastrado com sucesso' });
    });
    
    // Se chegou atÃ© aqui Ã© porque as operaÃ§Ãµes foram concluÃ­das com sucesso,
    // nÃ£o sendo necessÃ¡rio finalizar a transaÃ§Ã£o manualmente.
    // `result` terÃ¡ o resultado da transaÃ§Ã£o, no caso um empregado e o endereÃ§o cadastrado
  } catch (e) {
    // Se entrou nesse bloco Ã© porque alguma operaÃ§Ã£o falhou.
    // Nesse caso, o sequelize irÃ¡ reverter as operaÃ§Ãµes anteriores com a funÃ§Ã£o rollback, nÃ£o sendo necessÃ¡rio fazer manualmente
    console.log(e.message);
    res.status(500).json({ message: 'Algo deu errado' });
  }
});

app.get('/employees/:id', async (req, res) => {
  try {
    const { id } = req.params;
    // const employee = await Employee.findOne({
    //     where: { id },
          // Eager Loadingâ†´
    //     include: [{
    //       model: Address, as: 'addresses', attributes: { exclude: ['number'] },
    //     }],
    //   });

    const employee = await Employee.findOne({ where: { id } });

    if (!employee)
      return res.status(404).json({ message: 'FuncionÃ¡rio nÃ£o encontrado' });

    // // Lazy Loadingâ†´
    // if (req.query.includeAddresses === 'true') {
    //   const addresses = await Address.findAll({ where: { employeeId: id } });
    //   return res.status(200).json({ employee, addresses });
    // }

    if (!req.query.includeAddresses) return res.status(200).json(employee);

    const addresses = await employee.getAddresses(); //getter
    return res.status(200).json({ ...employee.dataValues, addresses})
  } catch (e) {
    console.log(e.message);
    res.status(500).json({ message: 'Algo deu errado' });
  };
});

app.get('/usersbooks/:id', async (req, res) => {
  try {
    const { id } = req.params;

    const user = await User.findOne({
      where: { userId: id },
      // Eager Loadingâ†´
      include: [{ model: Book, as: 'books', through: { attributes: [] } }],
    });

    if (!user)
      return res.status(404).json({ message: 'UsuÃ¡rio nÃ£o encontrado' });

    return res.status(200).json(user);
  } catch (e) {
    console.log(e.message);
    res.status(500).json({ message: 'Algo deu errado' });
  };
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => console.log(`Ouvindo na porta ${PORT}`));

module.exports = app;